### PR TITLE
bart summarization ORT+DeepSpeed extensions

### PR DESCRIPTION
**Pull request by jsleep**
* ORT+DeepSpeed extensions to BART Summarization example (thanks Han!)
* Improving the example code to make use of backend arg in trainer (specified in config), instead of handling it in the example code unnecessarily.

### CODE REVIEW

File: examples/summarization/deepspeed_methods/deepspeed_trainer_backend.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
+from tqdm import tqdm
+import torch
+from typing import Iterable, List, Optional, Union
+
+from pymarlin import SingleProcess
+from pymarlin.core import module_interface
+from pymarlin.core.trainer_backend import TrainerBackendArguments, OutputCollector, DDPTrainerBackend
+
+
+class deepspeed_trainer_backend(SingleProcess):
```
Comment (krishansubudhi) [inline examples/summarization/deepspeed_methods/deepspeed_trainer_backend.py:10]: DeepspeedBackend

File: examples/summarization/model_ortds.py [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+        self.DEEPSPEED_CKPT_PREFIX = "deepspeed_ckpt"
+    
+    @property
+    def pad_token_id(self):
+        return self._pad_token_id
+
+    def get_optimizers_schedulers(
+            self, estimated_global_steps_per_epoch: int, epochs: int
+    ):
+        if self.deepspeed:
```
Comment (krishansubudhi) [inline examples/summarization/model_ortds.py:72]: too many if statements for deepspeed. Possible refactorization using decorator pattern or multiple derived classes?

File: examples/summarization/train.py [hunk 3]
```diff
@@ -97,29 +103,36 @@ def collate_fun(self, batch):
...
+    def pad_token_id(self):
+        return self.model.config.pad_token_id
 
     def train_step(self, global_step: int, batch, device):
         batch = batch.to(device)
         result = self.model.forward(**batch)
         global_stats.update("lr", self.schedulers.get_last_lr()[0], frequent=True)
         return result.loss
+    
+    def get_core_model(self):
```
Comment (krishansubudhi) [inline examples/summarization/train.py:120]: can this be avoided? Looks like a redundant refatorization for deepspeed

File: examples/summarization/train.py [hunk 4]
```diff
@@ -3,10 +3,13 @@
 from pymarlin.core import trainer_backend, module_interface, trainer
 from transformers import BartForConditionalGeneration, BartTokenizerFast
 from torch.utils.data import DataLoader
+from deepspeed_methods.deepspeed_utils import get_core_model
```
Comment (krishansubudhi) [inline examples/summarization/train.py:6]: can we remove this from here?

File: examples/summarization/train.py [hunk 5]
```diff
@@ -180,8 +193,12 @@ def add_newline_to_end_of_each_sentence(x: str) -> str:
         else:
             return aggregator._scores  # here we return defaultdict(list)
 
+    @rank_zero_only
     def on_end_val_epoch(self, global_step, *collated_output, key="default"):
-        print('Evaluating gathered results.')
+        logger.info('Evaluating gathered results.')
+        if len(collated_output) == 0:
```
Comment (krishansubudhi) [inline examples/summarization/train.py:199]: Why is this required. Unnecessary error check ? We will always have more than zero validation examples collected

File: examples/summarization/train.py [hunk 6]
```diff
@@ -190,14 +207,13 @@ def on_end_val_epoch(self, global_step, *collated_output, key="default"):
         refs = self.tokenizer.batch_decode(
             labels, skip_special_tokens=True, clean_up_tokenization_spaces=False
         )
-        # print(refs[:2])
-        # print(preds[:2])
+        logger.info(f"preds[:2]: {preds[:2]}")
```
Comment (krishansubudhi) [inline examples/summarization/train.py:210]: please comment out debug code if it's not necessary

**Review by krishansubudhi — DISMISSED** [review]

File: examples/summarization/train.py (hunk 5 above)
Comment (jsleep) [reply examples/summarization/train.py:199]: this is something I think that was more required when Han was running and wasn't using @rank_zero_only, now that we are doing that I can get rid of this.

File: examples/summarization/train.py (hunk 6 above)
Comment (jsleep) [reply examples/summarization/train.py:210]: logger.debug() at least? log level will default to INFO anyways.

File: examples/summarization/model_ortds.py (hunk 2 above)
Comment (jsleep) [reply examples/summarization/model_ortds.py:72]: I can make a seperate module interface that can extend for deepspeed, and a decorator/wrapper class for ORT.

File: examples/summarization/train.py (hunk 3 above)
Comment (jsleep) [reply examples/summarization/train.py:120]: maybe - but i mostly did this so I don't have to copy paste the entire val_step again and instead just override get_core_model() which will vary for ORT/DS extension. I could do some fancy self.model property magic in ORT/Deepspeed extension but its just over kill for just wanting to get access to the original pytorch model after wrapping it twice for ORT/DS

Comment (jsleep) [reply examples/summarization/train.py:119]: leaving it in because it's a very small overridable function that's much better than copy-pasting the entire val_step across extended modules.

File: examples/summarization/model_ortds.py (hunk 2 above)
Comment (jsleep) [reply examples/summarization/model_ortds.py]: I cleaned these up and created an ORT and DeepSpeed module (DeepSpeed inherits from ORT for simplicity - but we could make these wrap for common trainers in the future).

**Review by ananthrs1 — DISMISSED** [review]

**Review by ananthrs1 — DISMISSED** [review]

**Review by ananthrs1 — APPROVED** [review]
